### DISCUSSION TITLE
how can I access the desterilized data returned from an azure table ?

### DISCUSSION THREAD

**Discussion by MIkey007007** [microsoft/BotFramework-Composer#5514] [Q&A]
I am using the composer to publish a bot to fetch data from an azure storage table.
The composer uses the drag and drops visual editor to generate the code using the REST API to fetch the data from the table.
Everything works fine until when I need to access the deserialized data returned by the storage rest API. The following is a copy of the deserialized data returned by the storage rest API:

<link rel="edit" title="mytable001" href="mytable001(PartitionKey='123',RowKey='123')" />
    <title />
    <updated>2021-01-13T17:56:26Z</updated>
    <author>
        <name />
    </author>
    <content type="application/xml">
        <m:properties>
            <d:PartitionKey>123</d:PartitionKey>
            <d:RowKey>123</d:RowKey>
            <d:Timestamp m:type="Edm.DateTime">2021-01-11T02:14:13.5692286Z</d:Timestamp>
            <d:name>deep learning</d:name>
            <d:url>https://en.wikipedia.org/wiki/Deep_learning</d:url>
        </m:properties>
    </content>
</entry>

In my code generated by the composer, the bot immediately does a "set property" step immediately following the successful return of the REST API (storage table query).   Given the deserialized object returned by the storage REST API, how should the "set property" statement be constructed so the bot can print our the individual data field,

Another way to phrase the question: how can I use the composer to construct the bot to iterate through a returned deserialized object (coded in XML JSON format)? 

Where can I find a document that can shed some light on this matter? 
Is there any place I can find a good example ?  Can it be done via composer?    

Thanks in advance,.

**Comment by dmvtech** [comment]
Hi @MIkey007007 

Can you clarify what you mean by XML JSON format? Usually these are two different data types. Are you talking about XML serialized into JSON, or the other way around? Or something that I am not familiar with (completely possible of course)? The example you have above:
`<title /> 2021-01-13T17:56:26Z 123 123 2021-01-11T02:14:13.5692286Z deep learning https://en.wikipedia.org/wiki/Deep_learning` is not valid (or at least well formed) JSON or XML.

> Another way to phrase the question: how can I use the composer to construct the bot to iterate through a returned deserialized object (coded in XML JSON format)?

Again, I'm not sure about the format you are talking about, but you can use the [foreach](https://docs.microsoft.com/en-us/azure/bot-service/adaptive-expressions/adaptive-expressions-prebuilt-functions?view=azure-bot-service-4.0#foreach) prebuilt function to iterate over an array or object.

**Reply by MIkey007007** [reply]
Hi @dmvtech,

Thanks for the response. My first question: we can always assume that the bot's interface (dialog.api_response.content) will always provide a "deserialized object". In another word, can we always assume that "dialog.api_response.content" will always provide an XML object?   Secondly, is there anywhere we can find a good example showing how we can iterate through the individual XML field returned by the dialog.api_response.content?  Perhaps I didn't put in the right search term, I couldn't find any good document specifying how to access the individual field returned by the dialog.api_response.content?   

Thanks in advance.

**Reply by dmvtech** [reply]
I responded to your stack overflow issue here:
https://stackoverflow.com/questions/65708302/how-can-i-compose-a-bot-to-iterate-trough-a-xml-json-object

But to clarify, you can use xPath. If it is more than a single value, it will return an array that you can iterate over that array for the values.

**Reply by dmvtech** [reply]
To summarize:

- Send http request can ask for either XML or JSON using the content type field
- Body can be supplied with a JSON object, just ensure object is set as type for Body field
- You can convert/query/deserialize using `=json()`, `=xPath()`, `=jPath()`, `=string()`
- You can use [type checking functions](https://docs.microsoft.com/en-us/azure/bot-service/adaptive-expressions/adaptive-expressions-prebuilt-functions?view=azure-bot-service-4.0#type-checking-functions) to check if is string or json.

**Reply by MIkey007007** [reply]
@dmvtech:

Is there anywhere I can find a good example showing how to query or iterate through a deserialized object returned by dialog.api_response.content?

Bot Composer document specifies that "If the response is JSON, it will be a deserialized object available via dialog.api_response.content". Yet, I couldn't find any example to elaborate on it.  thanks in advance. 

Can I do this????:  foreach (dilog.api_response.content))

**Reply by MIkey007007** [reply]
Thanks for the information.
